README.md
# TRNode
ode for three-node Transcriptional Regulatory Network

## HOW TO USE THIS PROGRAM ?

You can use this program under the terms of the GNU General Public License.

## HOW TO Execute ?

0. enter folder and press `make`.
1. single mode: `./trnode $network $parameter $etalist`.
2. loop mode: `./trnode $network_start $network_end`.
3. config mode: `./trnode $path_to_config_file`.

## HOW TO DEBUG using vscode ?

1. edit 4 path in `.vscode/launch.json` & 1 path in `tasks.json`.
2. select debug configuration to `trnode DBG`.
3. press `f5` to start.

## HOW TO CONTACT AUTHOR ?

email to `send2li@icloud.com`

## OPEN SOURCE

1. GNU Scientific Library.
2. Shi Wenjia

CHANGELOG.md
# CHANGELOG

## [3.1] - 2023-2-14

### Feature

* 1. add version control .git folder

### Fixed

* 1. Fix free operation in monitor function to avoid causing segment fault in the processor fetch behavior.
* 2. Fix incorrect calculation of progress and estimated time when encounter an unadaptable network.

### Refactored

* 1. Improve slot utilization by reducing malloc operation on different eta_qty.

4_ode.c
/* 4_ode.c
 *
 * Copyright (C) 2023 Song Changli
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or (at
 * your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 */

/*
 * Original Author: Brian Gough
 */

#include "0_header.h"

// 4th-order Rungekuta algorithm, which calculate a new step for special h
static void rk4 (_complex y[], _complex k[], double h, ode_struct* ode) {
    _complex y0[cst_node];
    _complex temp_y[cst_node];
    memcpy (y0, y, cst_node * sizeof (_complex));
    double c[2] = { global.cst_cal[2], global.cst_cal[3] };
    double h_half = h * 0.5;
    double hc0 = h * c[0];
    double hc1 = h * c[1];

    y[0] += hc0 * k[0];
    y[1] += hc0 * k[1];
    y[2] += hc0 * k[2]; // y1

    temp_y[0] = y0[0] + h_half * k[0];
    temp_y[1] = y0[1] + h_half * k[1];
    temp_y[2] = y0[2] + h_half * k[2];
    equation (k, temp_y, ode); // k2

    y[0] += hc1 * k[0];
    y[1] += hc1 * k[1];
    y[2] += hc1 * k[2]; // y2

    temp_y[0] = y0[0] + h_half * k[0];
    temp_y[1] = y0[1] + h_half * k[1];
    temp_y[2] = y0[2] + h_half * k[2];
    equation (k, temp_y, ode); // k3

    y[0] += hc1 * k[0];
    y[1] += hc1 * k[1];
    y[2] += hc1 * k[2]; // y3

    temp_y[0] = y0[0] + h * k[0];
    temp_y[1] = y0[1] + h * k[1];
    temp_y[2] = y0[2] + h * k[2];
    equation (k, temp_y, ode); // k4

    y[0] += hc0 * k[0];
    y[1] += hc0 * k[1];
    y[2] += hc0 * k[2]; // y4
}

// repeatly try a new step until satisfy error control condition
void step (ode_struct* ode, double* h, double* t) {
    _complex y_1step[cst_node];
    _complex y_2step[cst_node];
    _complex y_err[cst_node];
    _complex k[cst_node];
    _complex* dydt = ode->dydt;
    _complex* y0 = ode->y;
    double r[cst_node];
    double half_h;
    double h0 = *h;
    double h_old;
    double temp;
    double rmax;

    for (;;) {
        { // compare one whole step with two steps that devided into half
            memcpy (y_1step, y0, cst_node * sizeof (_complex));
            memcpy (k, dydt, cst_node * sizeof (_complex));
            rk4 (y_1step, k, h0, ode);

            half_h = 0.5 * h0;
            memcpy (y_2step, y0, cst_node * sizeof (_complex));
            memcpy (k, dydt, cst_node * sizeof (_complex));
            rk4 (y_2step, k, half_h, ode);
            equation (k, y_2step, ode);
            rk4 (y_2step, k, half_h, ode);

            equation (k, y_2step, ode);
            y_err[0] = global.cst_cal[4] * (y_2step[0] - y_1step[0]);
            y_err[1] = global.cst_cal[4] * (y_2step[1] - y_1step[1]);
            y_err[2] = global.cst_cal[4] * (y_2step[2] - y_1step[2]);
        }

        { // adjust h
            rmax = 0;
            h_old = h0;
            r[0] = fabs (y_err[0] / (eps_abs + eps_rel * y_2step[0]));
            r[1] = fabs (y_err[1] / (eps_abs + eps_rel * y_2step[1]));
            r[2] = fabs (y_err[2] / (eps_abs + eps_rel * y_2step[2]));
            rmax = r[0] > r[1] ? (r[0] > r[2] ? r[0] : r[2])
                               : (r[1] > r[2] ? r[1] : r[2]);
            if (rmax > 1.1) {
                temp = 0.9 * pow (rmax, -0.25);
                if (temp < 0.2)
                    temp = 0.2;
                h0 = temp * h_old;
                continue; // try with an updated h0
            } else if (rmax < 0.5) {
                temp = 0.9 * pow (rmax, -0.2);
                if (temp > 4.9)
                    temp = 4.9;
                else if (temp < 1.0)
                    temp = 1.0;
                h0 = temp * h_old; // increase the step with h0 next time and quit
            }

            *t += h_old;
            *h = h0;

            memcpy (y0, y_2step, cst_node * sizeof (_complex));
            memcpy (dydt, k, cst_node * sizeof (_complex));
            break;
        }
    }
}

1_fetch.c
/* 1_fetch.c
 *
 * Copyright (C) 2023 Song Changli
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or (at
 * your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 */

#include "0_header.h"
#include <fcntl.h>
#include <sys/stat.h>

global_struct global;

case_struct* CASE;

// check legality of program argument
static int check (int argc, char** argv, char* path) {
    FILE* fp;
    getcwd (global.path_env, FILENAME_MAX);
    char error_msg[FILENAME_MAX] = "\033[1m\033[31merror:\033[0m ";

    if (argc < 2 || argc > 6 || argc == 5) {
        printf ("%sargument incorrect.\n", error_msg);
        return 1;
    }
    sprintf (path, "%s/data/para/const", global.path_env);
    if (access (path, 0) || !(fp = fopen (path, "r"))) {
        printf ("%scan not open const file or path incorrect: %s.\n", error_msg, path);
        return 1;
    }
    fclose (fp);
    sprintf (path, "%s/data/para/LHS", global.path_env);
    if (access (path, 0) || !(fp = fopen (path, "r"))) {
        printf ("%scan not open LHS file or path incorrect: %s.\n", error_msg, path);
        return 1;
    }
    fclose (fp);

    switch (argc) {
    case 2:
        if (argv[1][0] == '/')
            sprintf (path, "%s", argv[1]);
        else
            sprintf (path, "%s/%s", global.path_env, argv[1]);
        if (access (path, 0) || !(fp = fopen (path, "r"))) {
            printf ("%scan not access config file or path incorrect: %s\n", error_msg, path);
            return 1;
        }
        fclose (fp);
        break;
    case 3:
        if (atoi (argv[1]) < 0 || atoi (argv[1]) >= 19683 || atoi (argv[2]) < 0 || atoi (argv[2]) >= 19683 || atoi (argv[1]) > atoi (argv[2])) {
            printf ("%sargument incorrect.\n", error_msg);
            return 1;
        }
        sprintf (path, "%s/data/para/eta", global.path_env);
        if (access (path, 0) || !(fp = fopen (path, "r"))) {
            printf ("%scan not access eta file or path incorrect: %s\n", error_msg, path);
            return 1;
        }
        fclose (fp);
        break;
    default:
        if (atoi (argv[1]) < 0 || atoi (argv[1]) >= 19683 || atoi (argv[2]) < 0 || atoi (argv[2]) >= cst_para) {
            printf ("%sargument incorrect.\n", error_msg);
            return 1;
        }
        break;
    }
    return 0;
}

// fetch 'const' and 'LHS' data then pre-proceed
static int data () {
    int up;
    FILE* fp;
    size_t i, j, k;
    char path[FILENAME_MAX];
    double kl, kh, nl, nh, taul, tauh, base, fold, tmp;

    sprintf (path, "%s/data/para/const", global.path_env);
    fp = fopen (path, "r");
    fscanf (fp, "%lf%lf%lf%lf%lf%lf%lf%lf%d", &kl, &kh, &nl, &nh, &taul, &tauh, &base, &fold, &up);
    fclose (fp);

    sprintf (path, "%s/data/para/LHS", global.path_env);
    fp = fopen (path, "r");
    for (i = 0; i < cst_para; ++i) {
        for (j = 0; j < cst_node; ++j)
            for (k = 0; k < cst_node; ++k) {
                fscanf (fp, "%lf", &tmp);
                global.kn[i][j][k] = kl * pow (10, tmp * log10 (kh / kl));
            }
        for (j = 0; j < cst_node; ++j)
            for (k = 0; k < cst_node; ++k) {
                fscanf (fp, "%lf", &tmp);
                global.n[i][j][k] = nl + tmp * (nh - nl);
                global.n2[i][j][k] = pow (global.n[i][j][k], 2);
                global.kn[i][j][k] = pow (1.0 / global.kn[i][j][k], global.n[i][j][k]);
            }
        for (j = 0; j < cst_node; ++j) {
            fscanf (fp, "%lf", &tmp);
            global.t1[i][j] = 1.0 / (taul * pow (10, tmp * log10 (tauh / taul)));
        }
    }
    fclose (fp);

    global.cst_cal[2] = 1.0 / 6.0;
    global.cst_cal[3] = 1.0 / 3.0;
    global.cst_cal[4] = 4.0 / 15.0;
    global.cst_cal[5] = 2 * eps_abs;
    global.cst_cal[6] = 2 * eps_abs + 1;
    global.cst_cal[0] = base / 0.4 / (1 + base / 0.4);
    base *= up * 10;
    global.cst_cal[1] = base / 0.4 / (1 + base / 0.4);

    return 0;
}

// fetch config file and generate case link list
static int cases (int argc, char** argv, int* eta_max, char* path) {
    case_struct* cur;
    size_t i, j, k;
    int consistent;
    int case_qty;
    double eta;
    FILE* fp;

    CASE = ( case_struct* )malloc (sizeof (case_struct));
    global.total = 0;
    *eta_max = 0;
    cur = CASE;
    switch (argc) {
    case 2: // config
        fp = fopen (path, "r");
        fscanf (fp, "%d", &case_qty);
        for (i = 0; i < case_qty; ++i) {
            cur->case_id = i;
            fscanf (fp, "%d %d", &consistent, &cur->eta_qty);
            *eta_max = cur->eta_qty > *eta_max ? cur->eta_qty : *eta_max;
            cur->eta_all = ( double** )malloc (cur->eta_qty * sizeof (double*));
            for (j = 0; j < cur->eta_qty; ++j) {
                cur->eta_all[j] = ( double* )malloc (cst_node * sizeof (double));
                if (consistent) {
                    fscanf (fp, "%lf", &eta);
                    for (k = 0; k < cst_node; ++k)
                        cur->eta_all[j][k] = eta;
                } else
                    fscanf (fp, "%lf %lf %lf", &cur->eta_all[j][0], &cur->eta_all[j][1], &cur->eta_all[j][2]);
            }
            fscanf (fp, "%d", &cur->content_qty);
            if (cur->content_qty == -1) { // config-loop
                cur->case_type = 1;
                cur->content_qty = 2;
                cur->content = ( int* )malloc (cur->content_qty * sizeof (int));
                fscanf (fp, "%d %d", &cur->content[0], &cur->content[1]);
                cur->progress[0] = cur->content[0];
                cur->progress[1] = 0;
                global.total += cst_para * (cur->content[1] - cur->content[0] + 1);
            } else { // discrete
                cur->case_type = 2;
                cur->content = ( int* )malloc (cur->content_qty * sizeof (int));
                for (j = 0; j < cur->content_qty; ++j)
                    fscanf (fp, "%d", &cur->content[j]);
                cur->progress[0] = 0;
                global.total += cur->content_qty;
            }
            if (i != case_qty - 1) {
                cur->next = ( case_struct* )malloc (sizeof (case_struct));
                cur = cur->next;
            }
            cur->next = NULL;
        }
        fclose (fp);
        return case_qty;
    case 3: // loop
        fp = fopen (path, "r");
        fscanf (fp, "%d %d", &consistent, &cur->eta_qty);
        *eta_max = cur->eta_qty > *eta_max ? cur->eta_qty : *eta_max;
        cur->eta_all = ( double** )malloc (cur->eta_qty * sizeof (double*));
        for (i = 0; i < cur->eta_qty; ++i) {
            cur->eta_all[i] = ( double* )malloc (cst_node * sizeof (double));
            if (consistent) {
                fscanf (fp, "%lf", &eta);
                for (j = 0; j < cst_node; ++j)
                    cur->eta_all[i][j] = eta;
            } else
                fscanf (fp, "%lf %lf %lf", &cur->eta_all[i][0], &cur->eta_all[i][1], &cur->eta_all[i][2]);
        }
        cur->case_id = 0;
        cur->next = NULL;
        cur->case_type = 1;
        cur->content_qty = 2;
        cur->content = ( int* )malloc (2 * sizeof (int));
        cur->content[1] = atoi (argv[2]);
        cur->progress[0] = cur->content[0] = atoi (argv[1]);
        global.total = cst_para * (cur->content[1] - cur->content[0] + 1);
        cur->progress[1] = 0;
        fclose (fp);
        return 1;
    default: // single
        global.total = 1;
        cur->case_id = 0;
        cur->next = NULL;
        cur->case_type = 0;
        cur->content_qty = 2;
        cur->eta_qty = argc - 3;
        *eta_max = cur->eta_qty;
        cur->content = ( int* )malloc (2 * sizeof (int));
        cur->eta_all = ( double** )malloc (cur->eta_qty * sizeof (double*));
        for (i = 0; i < cur->eta_qty; ++i) {
            cur->eta_all[i] = ( double* )malloc (cst_node * sizeof (double));
            for (j = 0; j < cst_node; ++j)
                cur->eta_all[i][j] = atof (argv[argc == 4 ? 3 : 3 + j]);
        }
        cur->content[0] = atoi (argv[1]);
        cur->content[1] = atoi (argv[2]);
        return 0;
    }
}

// create result folders and files base on current time and case quantity
static int files (int case_qty, char** argv) {
    char file[FILENAME_MAX], folder[FILENAME_MAX];
    struct tm* timer;
    size_t i;

    global.clock = time (NULL);
    timer = gmtime (&global.clock);
    sprintf (global.path_env, "%s/data/%d%02d%02d%02d%02d%02d", global.path_env, (1900 + timer->tm_year), (1 + timer->tm_mon), timer->tm_mday, timer->tm_hour, timer->tm_min, timer->tm_sec);
    if (access (global.path_env, 0))
        mkdir (global.path_env, 0777);

    if (case_qty)
        for (i = 0; i < case_qty; ++i) {
            sprintf (folder, "%s/%ld", global.path_env, i);
            if (access (folder, 0))
                mkdir (folder, 0777);
            sprintf (file, "%s/response", folder);
            if (access (file, 0))
                creat (file, 0777);
            sprintf (file, "%s/timeout", folder);
            if (access (file, 0))
                creat (file, 0777);
            sprintf (file, "%s/detail", folder);
            if (access (file, 0))
                creat (file, 0777);
        }
    else {
        sprintf (file, "%s/%09d", global.path_env, atoi (argv[1]) * 10000 + atoi (argv[2]));
        if (access (file, 0))
            creat (file, 777);
    }
    return 0;
}

// main procedure
int main (int argc, char* argv[]) {
    int eta_max;
    char path[FILENAME_MAX];
    if (check (argc, argv, path) || data () || files (cases (argc, argv, &eta_max, path), argv))
        exit (-1);

    init (CASE, eta_max, argc > 3);

    return 0;
}

0_header.h
/* 0_header.h
 *
 * Copyright (C) 2023 Song Changli
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or (at
 * your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 */

#include <complex.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#define eps_abs 1e-7
#define eps_rel 1e-7

#define mark_ntadapt -3
#define mark_timeout -2
#define mark_outlier -1

#define cst_node 3
#define cst_para 10000

typedef double complex _complex;

typedef struct {
    double kn[cst_para][cst_node][cst_node]; // k^(-n)
    double n2[cst_para][cst_node][cst_node]; // n^2
    double n[cst_para][cst_node][cst_node];
    double t1[cst_para][cst_node]; // tau^(-1)

    char path_env[FILENAME_MAX];
    double cst_cal[7]; // constant: input, rk4, eps
    time_t clock;
    int total;
} global_struct;

typedef struct case_struct {
    int case_id;
    int eta_qty;
    int* content;  // single: net, para; loop: net_start, net_end; discret: uid
    int case_type; // 0: single; 1: loop; 2: discrete
    int content_qty;
    int progress[2]; // loop: net_i, para_i; discrete: case_i
    double** eta_all;
    struct case_struct* next;
} case_struct;

typedef struct frame_struct {
    _complex dydt[cst_node];
    _complex y[cst_node];
    double t;
    int phase;
    int flag_stable;
    int flag_outlier;
    int time_consume;
    struct frame_struct* next;
} frame_struct;

typedef struct {
    _complex y[cst_node];
    _complex dydt[cst_node];
    _complex cst_top[cst_node];
    double eta_list[cst_node];
    int net_top[cst_node][cst_node];
    int flag_record;
    size_t para;
} ode_struct;

extern global_struct global;

double inspect (double*, frame_struct*, ode_struct*);
void equation (_complex*, _complex*, ode_struct*);
void step (ode_struct*, double*, double*);
void init (case_struct*, int, int);

5_equation.c
/* 5_equation.c
 *
 * Copyright (C) 2023 Song Changli
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or (at
 * your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 */

#include "0_header.h"

// calculate TRN dydt by provided ode infomation
void equation (_complex dydt[], _complex y[], ode_struct* ode) {
    size_t i, j, para;
    register _complex reg;
    _complex tmp, retro[cst_node] = { 0, 0, 0 };

    para = ode->para;
    double* t1 = global.t1[para];
    double* eta_list = ode->eta_list;
    double (*n)[cst_node] = global.n[para];
    double (*n2)[cst_node] = global.n2[para];
    double (*kn)[cst_node] = global.kn[para];
    int (*net_top)[cst_node] = ode->net_top;
    memcpy (dydt, ode->cst_top, cst_node * sizeof (_complex));

    for (i = 0; i < cst_node; ++i)
        for (j = 0; j < cst_node; ++j)
            if (net_top[i][j]) {
                reg = cpow (y[i], n[i][j] - 1);         // y^(n-1)
                tmp = reg * n2[i][j] * kn[i][j];        // n^2*y^(n-1)*k^(-n)
                reg *= y[i] * kn[i][j];                 // (y/k)^n
                dydt[j] *= net_top[i][j] - 1 ? 1 : reg; // node_i activate node_j
                reg = 1.0 / ++reg;                      // 1/(1+(y/k)^n)
                dydt[j] *= reg;                         // (y/k)^n/(1+(y/k)^n)
                retro[i] += tmp * reg * reg;            // n^2*y^(n-1)*k^(-n)/(1+(y/k)^n)^2
            }

    dydt[0] = t1[0] * (dydt[0] - y[0]) / (1 + retro[0] * eta_list[0]);
    dydt[1] = t1[1] * (dydt[1] - y[1]) / (1 + retro[1] * eta_list[1]);
    dydt[2] = t1[2] * (dydt[2] - y[2]) / (1 + retro[2] * eta_list[2]);
}

2_queue.c
/* 2_queue.c
 *
 * Copyright (C) 2023 Song Changli
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or (at
 * your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 */

#include "0_header.h"
#include <pthread.h>

#define QUEUE_MAX_LENGTH 10000

typedef struct {
    ode_struct ode;
    frame_struct* frame;
    double** eta_all;
    double** result;
    int case_type;
    int case_id;
    int eta_qty;
    int status; // 0.unsolve 1.solved 2.done
    int uid;
} slot_struct;

static struct {
    pthread_mutex_t lock;
    slot_struct* slot;
    size_t front, rear;
} QUEUE;

static case_struct* CASE;

// generate and judge whether net topology can possibly achieve adaptation
static int topology (int (*net_top)[cst_node], int flag_hill[], int net) {
    size_t i, j;
    int tmp = net * 3;
    for (i = 0; i < cst_node; ++i)
        for (j = 0; j < cst_node; ++j)
            net_top[i][j] = (tmp /= 3) % 3 - 1;
    if (fabs (net_top[0][1]) + fabs (net_top[0][2]) + fabs (net_top[1][0]) + fabs (net_top[1][2]) + fabs (net_top[2][0]) + fabs (net_top[2][1]) <= 1 || !(net_top[0][2] || (net_top[0][1] && net_top[1][2])))
        return 0;
    for (i = 1; i < cst_node; ++i) {
        flag_hill[i - 1] = 0;
        for (j = 0; j < cst_node; ++j)
            flag_hill[i - 1] = net_top[j][i] ? 1 : flag_hill[i - 1];
    }
    return 1;
}

// output completed task to result file
static int depart () {
    char status[4][FILENAME_MAX] = { "outlier", "timeout", "ntadapt", "\033[1m\033[32madapt\033[0m" };
    char path[2][FILENAME_MAX];
    double **datas, *data;
    int flag_output;
    size_t i, j;
    frame_struct *frame, *frame_pre;
    slot_struct* slot = &QUEUE.slot[QUEUE.rear];
    FILE* fp[2];

    if (slot->status != 1)
        return slot->status;

    if (slot->ode.flag_record) {
        data = slot->result[0];
        sprintf (path[0], "%s/%09d", global.path_env, slot->uid);
        fp[0] = fopen (path[0], "w+");
        frame = slot->frame->next;
        slot->frame->next = NULL;
        while (frame) {
            fprintf (fp[0], " %12lf %12lf %12lf %12lf\n", frame->t, creal (frame->y[0]), creal (frame->y[1]), creal (frame->y[2]));
            frame_pre = frame;
            frame = frame->next;
            free (frame_pre);
        }

        printf ("> %s\n", data[0] < 0 ? status[( int )fabs (data[0]) - 1] : status[3]);
        if (data[0] > 0 || data[0] == mark_ntadapt)
            printf ("response\t%12lf\npeakfold\t%12lf\nerror\t\t%12lf\nypeak\t\t%12lf\nystable\t\t%12lf\t%12lf\t%12lf\n\t\t%12lf\t%12lf\t%12lf\n", data[0], data[1], data[2], data[3], data[4], data[5], data[6], data[7], data[8], data[9]);
    } else {
        datas = slot->result;
        if (datas[0][0] == mark_timeout) {
            sprintf (path[0], "%s/%d/timeout", global.path_env, slot->case_id);
            fp[0] = fopen (path[0], "a+");
            fprintf (fp[0], "%09d\n", slot->uid);
            fclose (fp[0]);
        } else {
            flag_output = 0;
            for (i = 0; !flag_output && i < slot->eta_qty; ++i)
                flag_output = datas[i][0] > 0;
            if (flag_output) {
                sprintf (path[0], "%s/%d/response", global.path_env, slot->case_id);
                sprintf (path[1], "%s/%d/detail", global.path_env, slot->case_id);
                fp[0] = fopen (path[0], "a+");
                fp[1] = fopen (path[1], "a+");

                fprintf (fp[0], "%09d", slot->uid);
                for (i = 0; i < slot->eta_qty; ++i)
                    if (datas[i][0] > 0)
                        fprintf (fp[0], "%10.6lf", datas[i][0]);
                    else
                        fprintf (fp[0], "%10d", ( int )datas[i][0]);
                fprintf (fp[0], "\n");

                for (i = 0; i < slot->eta_qty; ++i) {
                    fprintf (fp[1], "%09d %6.2lf %6.2lf %6.2lf", slot->uid, slot->eta_all[i][0], slot->eta_all[i][1], slot->eta_all[i][2]);
                    if (datas[i][0] > 0)
                        fprintf (fp[1], "%10.6lf", datas[i][0]);
                    else
                        fprintf (fp[1], "%10d", ( int )datas[i][0]);
                    for (j = 1; j < 10; ++j)
                        fprintf (fp[1], " %10.6lf", datas[i][j]);
                    fprintf (fp[1], "\n");
                }
                fclose (fp[0]);
                fclose (fp[1]);
            }
        }
    }

    return 1;
}

// assign new task then drop into empty slot
static void enter () {
    int i;
    int is_pass = 0;
    int flag_hill[cst_node - 1];
    case_struct* case_pre;
    _complex y0[cst_node] = { 0.2, 0.2, 0.2 };
    slot_struct* slot = &QUEUE.slot[QUEUE.rear];

    slot->status = 2;
    while (CASE && slot->status == 2) {
        switch (CASE->case_type) {
        case 0:
            if (!topology (slot->ode.net_top, flag_hill, CASE->content[0])) {
                free (CASE->content);
                CASE = CASE->next;
                --global.total;
                break;
            }

            slot->status = 0;
            slot->ode.flag_record = 1;
            slot->case_id = CASE->case_id;
            slot->eta_qty = CASE->eta_qty;
            slot->case_type = CASE->case_type;
            slot->ode.para = CASE->content[1];
            slot->ode.cst_top[1] = ( _complex )flag_hill[0];
            slot->ode.cst_top[2] = ( _complex )flag_hill[1];
            slot->uid = CASE->content[0] * 10000 + (CASE->content[1]);
            memcpy (slot->ode.y, y0, cst_node * sizeof (_complex));
            for (i = 0; i < slot->eta_qty; ++i)
                slot->eta_all[i] = CASE->eta_all[i];
            slot->frame->next = NULL;
            CASE = CASE->next;
            break;
        case 1:
            if (CASE->progress[0] > CASE->content[1]) {
                free (CASE->content);
                CASE = CASE->next;
                break;
            }
            if (CASE->progress[1] >= cst_para || (is_pass = !topology (slot->ode.net_top, flag_hill, CASE->progress[0]))) {
                ++CASE->progress[0];
                CASE->progress[1] = 0;
                global.total -= is_pass ? cst_para : 0;
                break;
            }

            slot->status = 0;
            slot->ode.flag_record = 0;
            slot->case_id = CASE->case_id;
            slot->eta_qty = CASE->eta_qty;
            slot->case_type = CASE->case_type;
            slot->ode.para = CASE->progress[1];
            slot->ode.cst_top[1] = ( _complex )flag_hill[0];
            slot->ode.cst_top[2] = ( _complex )flag_hill[1];
            slot->uid = CASE->progress[0] * 10000 + (CASE->progress[1]);
            memcpy (slot->ode.y, y0, cst_node * sizeof (_complex));
            for (i = 0; i < slot->eta_qty; ++i)
                slot->eta_all[i] = CASE->eta_all[i];
            slot->frame->next = NULL;
            ++CASE->progress[1];
            break;
        case 2:
            if (CASE->progress[0] >= CASE->content_qty) {
                free (CASE->content);
                CASE = CASE->next;
            }
            if (!topology (slot->ode.net_top, flag_hill, CASE->content[CASE->progress[0]] / 10000)) {
                ++CASE->progress[0];
                --global.total;
                break;
            }

            slot->status = 0;
            slot->ode.flag_record = 0;
            slot->case_id = CASE->case_id;
            slot->eta_qty = CASE->eta_qty;
            slot->case_type = CASE->case_type;
            slot->uid = CASE->content[CASE->progress[0]];
            slot->ode.cst_top[1] = ( _complex )flag_hill[0];
            slot->ode.cst_top[2] = ( _complex )flag_hill[1];
            memcpy (slot->ode.y, y0, cst_node * sizeof (_complex));
            slot->ode.para = CASE->content[CASE->progress[0]] % 10000;
            for (i = 0; i < slot->eta_qty; ++i)
                slot->eta_all[i] = CASE->eta_all[i];
            slot->frame->next = NULL;
            ++CASE->progress[0];
            break;
        }
    }
}

// monitor task loop queue, output estimate remain time
static void monitor (int flag_record) {
    time_t clock = global.clock;
    int done = 0;

    for (;;) {
        switch (depart ()) {
        case 2:
            return;
        case 1:
            ++done;
            enter ();
            QUEUE.rear = (QUEUE.rear + 1) % QUEUE_MAX_LENGTH;
            if (!flag_record && global.total)
                printf ("\rprogress: %3d %%, estimate time remain: %.0f sec ", done * 100 / global.total, (global.total - done) * difftime (time (NULL), clock) / done);
        default:
            break;
        }
    }
}

// processor for calculating model
static void* processor () {
    size_t i, fetch, *front = &QUEUE.front;
    slot_struct* slot;

    for (;;) {
        fetch = -1;
        pthread_mutex_lock (&QUEUE.lock);
        fetch = *front;
        *front = (*front + 1) % QUEUE_MAX_LENGTH;
        pthread_mutex_unlock (&QUEUE.lock);
        if (!(fetch + 1))
            continue;
        slot = &QUEUE.slot[fetch];
        while (slot->status == 1)
            ;
        if (slot->status == 2)
            break;
        for (i = 0; i < slot->eta_qty; ++i) {
            memcpy (slot->ode.eta_list, slot->eta_all[i], cst_node * sizeof (double));
            if (inspect (slot->result[i], slot->frame, &slot->ode) == mark_timeout) {
                slot->result[0][0] = mark_timeout;
                break;
            }
        }
        slot->status = 1;
    }
    return ( void* )slot;
}

// initialize task loop queue
void init (case_struct* head, int eta_max, int flag_record) {
    int thread_qty = sysconf (_SC_NPROCESSORS_ONLN);
    pthread_t pid[thread_qty];
    case_struct* case_pre;
    size_t i, j;

    CASE = head;
    QUEUE.front = QUEUE.rear = 0;
    pthread_mutex_init (&QUEUE.lock, NULL);
    QUEUE.slot = ( slot_struct* )malloc (QUEUE_MAX_LENGTH * sizeof (slot_struct));
    for (i = 0; i < QUEUE_MAX_LENGTH; ++i) {
        slot_struct* slot = &QUEUE.slot[i];
        slot->status = 2;
        slot->frame = ( frame_struct* )malloc (sizeof (frame_struct));
        slot->result = ( double** )malloc (eta_max * sizeof (double*));
        slot->eta_all = ( double** )malloc (eta_max * sizeof (double*));
        for (j = 0; j < eta_max; ++j)
            slot->result[j] = ( double* )malloc (10 * sizeof (double));
        slot->uid = slot->eta_qty = slot->case_id = slot->case_type = slot->ode.para = slot->ode.flag_record = 0;
        CASE ? enter () : 1;
        QUEUE.rear = (QUEUE.rear + 1) % QUEUE_MAX_LENGTH;
    }

    QUEUE.rear = 0;
    for (i = 0; i < thread_qty; ++i)
        pthread_create (&pid[i], NULL, processor, NULL);
    monitor (flag_record);

    for (i = 0; i < thread_qty; ++i)
        pthread_join (pid[i], NULL);

    // free malloc space
    while (head) {
        for (i = 0; i < eta_max; ++i)
            free (head->eta_all[i]);
        free (head->eta_all);
        case_pre = head;
        head = head->next;
        free (case_pre);
    }
    for (i = 0; i < QUEUE_MAX_LENGTH; ++i) {
        for (j = 0; j < eta_max; ++j)
            free (QUEUE.slot[i].result[j]);
        free (QUEUE.slot[i].result);
        free (QUEUE.slot[i].eta_all);
    }
    free (QUEUE.slot);
    printf ("\r%s", flag_record ? "" : "\ndone.\n");
}